lib/src/rules/avoid_types_on_closure_parameters.dart

```diff
@@ -86,12 +85,42 @@ class AvoidTypesOnClosureParametersVisitor extends SimpleAstVisitor {
 }
 
 class _Visitor extends SimpleAstVisitor<void> {
+  _Visitor(this.rule, this.context);
   final LintRule rule;
-
-  _Visitor(this.rule);
+  final LinterContext context;
 
   @override
   void visitFunctionExpression(FunctionExpression node) {
+    var staticParameterElement = node.staticParameterElement;
+    var parent = node.parent;
+    while (parent is ParenthesizedExpression) {
+      parent = parent.parent;
+    }
+    DartType? expectedType;
+    if (parent is AssignmentExpression) {
+      expectedType = parent.writeType;
+    } else if (parent is VariableDeclaration) {
+      var parentParent = parent.parent;
+      if (parentParent is VariableDeclarationList) {
+        if (parentParent.type == null) {
+          // type infered: allow types
+          return;
+        }
+      }
+      expectedType = parent.declaredElement?.type;
+    } else if (parent is ReturnStatement || parent is ExpressionFunctionBody) {
+      expectedType =
+          node.thisOrAncestorOfType<FunctionDeclaration>()?.returnType?.type;
+    } else if (staticParameterElement != null) {
+      expectedType = staticParameterElement.type;
+    } else if (parent is NamedExpression) {
+      expectedType = parent.element?.type;
```

Comment:
@bwilkerson I was surprise to hava to add this line (to catch the `Future.then(..., onError: ...)` case). I thought `staticParameterElement` wouldn't have been null here. Is it the normal behaviour or a bug?

Reply:
It does seem inconsistent that an expression used as an argument wouldn't have a `staticParameterElement`. I'm not sure whether there's a reason for the current behavior, but I'd like to explore changing it. We should probably open an issue in the sdk repo and loop Konstantin into the conversation.

Reply:
Looks like this bug was already reported (by me :D) in https://github.com/dart-lang/sdk/issues/45964